webapp/graphite/render/datalib.py

```diff
@@ -243,3 +261,51 @@ def nonempty(series):
       return True
 
   return False
+
+
+class PrefetchedData(Future):
+  def __init__(self, results):
+    self._results = results
+    self._prefetched = None
+
+  def _data(self):
+    if self._prefetched is None:
+      self._fetch_data()
+    return self._prefetched
+
+  def _fetch_data(self):
+    prefetched = collections.defaultdict(list)
+    for result in self._results:
+      fetched = wait_for_result(result)
+
+      if fetched is None:
+        continue
+
+      for result in fetched:
+        prefetched[result['pathExpression']].append((
+          result['name'],
+          (
+            result['time_info'],
+            result['values'],
+          ),
+        ))
+
+    self._prefetched = prefetched
+
+
+def prefetchRemoteData(requestContext, targets):
+  """Prefetch a bunch of path expressions and stores them in the context.
+
+  The idea is that this will allow more batching that doing a query
+  each time evaluateTarget() needs to fetch a path. All the prefetched
+  data is stored in the requestContext, to be accessed later by datalib.
+  """
+  log.rendering("Prefetching remote data")
+  pathExpressions = extractPathExpressions(targets)
+
+  time_info = timebounds(requestContext)
+  (startTime, endTime, now) = time_info
+
+  results = STORE.fetch_remote(pathExpressions, startTime, endTime, now, requestContext)
+
+  requestContext['prefetched'][time_info] = PrefetchedData(results)
```

Comment:
We should be careful here, since when we pull the data back out we use the explicit `prefetched = requestContext['prefetched'].get((startTime, endTime, now), None)`, if `timebounds()` is updated to return any additional items then the lookup will fail.  This should likely be `requestContext['prefetched'][(startTime, endTime, now)] = PrefetchedData(results)`